### PR TITLE
feat(experimental): Add client library API to list recently updated jobs

### PR DESCRIPTION
**Pull request by mwiebe**
### What was the problem/requirement? (What/Why)

For the incremental downloads CLI command in development, we need a robust way to get
all jobs that are new, or have had tasks requeued, since a specified timestamp.

### What was the solution? (How)

* Added experimental candidate API function in deadline.client.api._list_jobs_recent_by_timestamp_field
  * It returns all jobs in a queue whose selected timestamp field is >= a provided timestamp, or since a provided timedelta ago.
  * The name starts with list_jobs_, so that someone using tab-complete or browsing the list of functions sees it alphabetically adjacent to the straight list_jobs function.
* The purpose of this function is to support the new incremental downloads CLI command. We expect it to be a useful primitive, but are marking it internal-only for now.

### What is the impact of this change?

We can build the rest of the CLI command using this function.

### How was this change tested?

Added unit tests that cover the function by generating random fake job datasets, and a mocked SearchJobs implementation that implements the subset of that API used by the function.

Tested the function on a Deadline Cloud queue.

### Was this change documented?

No.

### Does this PR introduce new dependencies?

No.

### Is this a breaking change?

No.

### Does this change impact security?

No.

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+            or look_back_duration must be provided.
+      look_back_duration (Optional[timedelta]): A timedelta subtracted from the current time to
+            get the timestamp from which to retrieve jobs. Either timestamp or look_back_duration
+            must be provided.
+
+    Returns:
+      The list of all jobs in the queue whose timestamp field is equal or greater than the provided
+      timestamp. Each job is as returned by the deadline:SearchJobs API.
+    """
+    # This function gets the full set of jobs whose timestamp field is newer than the thresho9ld timestamp.
```
Comment (leongdl) [inline src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py]: `thresho9ld` :)

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            "createdAt",
+            look_back_duration=timedelta(minutes=5)
+      )
+
+    Args:
+      boto3_session (boto3.Session): The boto3 Session for AWS API access.
+      farm_id (str): The Farm ID.
+      queue_id (str): The Queue ID.
+      timestamp_field_name (str): The name of the job timestamp field to use. Can be
+            "createdAt", "endedAt", "startedAt", or "updatedAt". See
```
Comment (leongdl) [inline src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:91]: Are we still considering `updatedAt` considering the corner cases we found?

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            queue_id,
+            "createdAt",
+            look_back_duration=timedelta(minutes=5)
+      )
+
+    Args:
+      boto3_session (boto3.Session): The boto3 Session for AWS API access.
+      farm_id (str): The Farm ID.
+      queue_id (str): The Queue ID.
+      timestamp_field_name (str): The name of the job timestamp field to use. Can be
```
Comment (leongdl) [inline src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:90]: We assume `timestamp_field_name` to be correct, do we want to guard rail this function with the enum set of `createdAt`, `endedAt`, `startedAt`, `updatedAt`, especially if we are going to make this a proper API?

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            # Rare edge case where the timestamp field is the same for all 100 jobs in the page, that
+            # we expect to never see in practice. The timestamp value is stored with
+            # millisecond precision, and jobs are scheduled independently from each other,
+            # with updates of running jobs generally being multiple seconds apart.
+            raise JobFetchFailure(
+                f"Failure fetching recent jobs based on the {timestamp_field_name} field as more then 100 jobs have the exact same timestamp value."
+            )
+        else:
+            # Continue processing from the largest timestamp value we saw so far
+            threshold_timestamp = jobs[-1][timestamp_field_name]
```
Comment (leongdl) [inline src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:210]: I think there's a case to considering here I was thinking over.

Consider a farm with 99 jobs already existing. We can call them [ j1 .... j99 ]. Now consider the user submits 2 jobs, j100, j101 in this order with sequential created at. And pretend the job number is also T=n

If this tool runs when the following sets are eventually consistent we may have a corner case if `j101` is indexed or returns earlier than `j100`.  `[ j1.... j99,  j101 ]`, leading to 1 page of data, meeting the `len(jobs) == total_results` exit condition. 

The next time this tool runs, and all 101 jobs are now indexed, we have a condition to consider. Would `threshold_timestamp` be set to the prior value of `j101` ? If yes, then the `filter_expression` will no longer include `j100`

Will we need to do an overlap algorithm? sort of like a time based ladder?

More so thinking of correctness if we lost 1 job that we did not download.

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py (hunk 2 above)
Comment (mwiebe) [reply src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:91]: From the perspective of this primitive function, yes. The corner cases were about how we tried to apply the updatedAt timestamp.

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py (hunk 4 above)
Comment (mwiebe) [reply src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:210]: To confirm what you're describing, do you mean the following?

What do we do if the assumption about system behavior in this part of the proof is wrong:

2. If j has the timestamp field created or modified during the algorithm, its value will be set to the current time
      and it will appear in a later set J_n, which is the last one at the time of create/update, but may not be
      the last time when the algorithm completes.

In particular, the trouble is when jobs get added/updated in the SearchJobs materialized view in a different order compared to the timestamp.

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    Failure fetching jobs updated since timestamp
+    """
+
+
+def _list_jobs_recent_by_timestamp_field(
+    boto3_session: boto3.Session,
+    farm_id: str,
+    queue_id: str,
+    timestamp_field_name: str,
+    *,
```
Comment (sakshie95) [inline src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:27]: Why do we need this?

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    timestamp: Optional[datetime] = None,
+    look_back_duration: Optional[timedelta] = None,
```
Comment (sakshie95) [inline src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:29]: Thanks for incorporating this over the union :) It's crystal clear to use as a library now

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+) -> list[dict[str, Any]]:
```
Comment (sakshie95) [inline src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:30]: Should we model this into a ``Job`` class instead to make this strongly typed?

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    #
+    # To prove these sets are equal, we show they are each a subset of the other.
+    #
+    # J' is a subset of J, because each J_i is from a deadline:SearchJobs that filters to a subset of J.
+    #
+    # Consider a job j ∈ J. A property of the system is that if the timestamp field is updated, it will be updated to the current time.
+    #    1. If the timestamp field of j is not updated during the algorithm, it will be
+    #       in one of the sets J_i because by construction the union of the query time intervals
+    #       cover the time interval for J.
+    #    2. If j has the timestamp field created or modified during the algorithm, its value will be set to the current time
```
Comment (sakshie95) [inline src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py]: This actually depends on the field they choose in the ``timestamp_field_name``. For eg. a created timestamp is never modified.. an updated timestamp is currently never modified except when the customer calls ``update_job``. Should we limit this algorithm to permit the timestamp fields that make a 100% sense for this algorithm to work as expected instead of the entire set of supported timestamps on deadline? i.e the variable ``permitted_field_names`` below

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        # Some jobs may not have the requested time stamp field.
+        jobs = [job for job in jobs if timestamp_field_name in job]
+
+        result_jobs.update({job["jobId"]: job for job in jobs})
+
+        if len(jobs) == total_results:
+            # If the jobs we got are the total results, result_jobs is now the full set
+            break
+        elif jobs[0][timestamp_field_name] == jobs[-1][timestamp_field_name]:
+            # Rare edge case where the timestamp field is the same for all 100 jobs in the page, that
```
Comment (sakshie95) [inline src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:201]: Maybe this should be in the library's docstring instead given that it completely fails the algorithm - so it's a callout when someone tries to use this library for their use-case

File: test/unit/deadline_client/api/test_list_jobs_recent_by_timestamp_field.py [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+    def _fake_search_jobs(
+        farmId, queueIds, itemOffset, pageSize=100, filterExpressions={}, sortExpressions=[]
+    ):
+        # Make deep copies of these parameters so we can destructively assert about them
+        filterExpressions = deepcopy(filterExpressions)
+        sortExpressions = deepcopy(sortExpressions)
+
+        assert farmId == farmIdForJobs
+        assert queueIdForJobs in queueIds
+        # We're only implementing ascending order by updatedAt
```
Comment (sakshie95) [inline test/unit/deadline_client/api/test_list_jobs_recent_by_timestamp_field.py]: Is this accurate? We've learnt updatedAt isn't updated for operations other than customer actions on resources so I'm not sure how this would help?

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+            # we expect to never see in practice. The timestamp value is stored with
+            # millisecond precision, and jobs are scheduled independently from each other,
+            # with updates of running jobs generally being multiple seconds apart.
+            raise JobFetchFailure(
+                f"Failure fetching recent jobs based on the {timestamp_field_name} field as more then 100 jobs have the exact same timestamp value."
```
Comment (benl-2023) [inline src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:206]: Could we shrink the probability we hit this edge case by increasing the pageSize used in the search call from the default to say 1000?

**Comment by sakshie95** [comment]
SonarCloud analysis is pretty extra strict about duplicated lines even in tests :) You might wanna make some param lists common in the test functions looking at the report https://sonarcloud.io/component_measures?id=aws-deadline_deadline-cloud&pullRequest=717&metric=new_duplicated_lines_density&view=list

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py (hunk 2 above)
Comment (leongdl) [reply src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:91]: Sure - more of a consumer question if it gets raised. Maybe a site note to add of the behaviour where it works, or where it does not.

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py (hunk 4 above)
Comment (leongdl) [reply src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:210]: > In particular, the trouble is when jobs get added/updated in the SearchJobs materialized view in a different order compared to the timestamp.

Effectively yes. Given the nature of eventual consistency and of search (and DDB), it is likely in a sequence of jobs items will be returned out of order, and therefore exiting the algorithm. The timestamp of each job is not controlled by this algorithm, and regardless of whether it is updated, or created during the algorithm the assumption is the result set can be unstable.

File: test/unit/deadline_client/api/test_list_jobs_recent_by_timestamp_field.py (hunk 10 above)
Comment (mwiebe) [reply test/unit/deadline_client/api/test_list_jobs_recent_by_timestamp_field.py]: Yep, it's a stale comment. Updating it to reflect that this fake function does ascending order by any of the timestamp fields.

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py (hunk 5 above)
Comment (mwiebe) [reply src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:27]: This makes `timestamp` and `look_back_duration` into keyword-only arguments. Because you can't specify both, it's awkward to try using them as positional. See https://peps.python.org/pep-3102/ about keyword-only arguments in general (this PEP is nice and short).

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py (hunk 6 above)
Comment (mwiebe) [reply src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:29]: Thanks for the input!

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py (hunk 11 above)
Comment (mwiebe) [reply src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:206]: From my analysis in the comment, the probability of this is already vanishingly small. It would be great to do some stress tests to double check that theory, since there could be factors I didn't account for in that.

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py (hunk 9 above)
Comment (mwiebe) [reply src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:201]: Yeah, that makes sense. I don't expect this to ever happen in practice, so I would put it as a footnote there.

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py (hunk 8 above)
Comment (mwiebe) [reply src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py]: I don't believe we need to change anything because the logic still applies to the fields that are never modified. In those cases we could simplify the statement to:

> If j has the timestamp field created during the algorithm, ...

By keeping the statement more generally, the reasoning applies to all the timestamps.

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py (hunk 7 above)
Comment (mwiebe) [reply src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:30]: The reason I avoided that is because the functions in this api namespace are structured to be like boto3 but with a little bit more convenience. We don't have a full wrapper of the Deadline Cloud APIs. If we modeled it as a `Job` class, then code that uses these functions and boto3 directly won't be as consistent.

**Comment by mwiebe** [comment]
> SonarCloud analysis is pretty extra strict about duplicated lines even in tests :) You might wanna make some param lists common in the test functions looking at the report https://sonarcloud.io/component_measures?id=aws-deadline_deadline-cloud&pullRequest=717&metric=new_duplicated_lines_density&view=list

Yeah SonarCloud is picky! We haven't marked it as required in github, so we can apply our judgment about how much to satisfy what it suggests. I try to think through "is this or that more readable" rather than worry specifically about duplication. Often too much duplicated code is unreadable, so it does reduce that, but not always. Sometimes duplicated code is better, because deduplication would move the lines of code you need to see far apart in the file. I'll see if I find a way to remove that duplication but not harm readability.

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py (hunk 2 above)
Comment (mwiebe) [reply src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:91]: We need to update https://docs.aws.amazon.com/deadline-cloud/latest/APIReference/API_GetJob.html#deadlinecloud-GetJob-response-updatedAt to be more clear.

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py (hunk 4 above)
Comment (mwiebe) [reply src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:210]: Following some higher bandwidth conversation, I've updated the docstring and the reasoning about correctness to include eventual consistency. What I've tried to do is document clearly what this function provides, and provide a generous overlap window in an example of 90 seconds, so that someone using it can reason appropriately.

**Review by leongdl — APPROVED** [review]
Thank you for the deep dive into the correctness of this function!

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py (hunk 7 above)
Comment (sakshie95) [reply src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:30]: interesting.. I would think we should have a full wrapper of the deadline cloud apis in that case for consistency but that's probably out of the scope of this PR

**Review by sakshie95 — APPROVED** [review]

File: src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py (hunk 7 above)
Comment (mwiebe) [reply src/deadline/client/api/_list_jobs_recent_by_timestamp_field.py:30]: Definitely something to consider! When I set this up, I decided against the idea because boto3 is the standard way to access AWS APIs from Python, and I tried to find the least-effort way to provide some patterns around using the APIs when logged in from Deadline Cloud Monitor vs with other AWS credentials.
